### PR TITLE
Tx queue, transaction confirmation

### PR DESCRIPTION
**Pull request by stackchain**
- Added transaction queue logic, rely on `isQueueOnline` from server status (not implemented yet)
- If the queue is offline the process follows as it is today
- Updated Delegation/Catalyst/Send screens to have the same logic while sending transactions
- It is possible to cancel an iteration with a ledger wallet

### CODE REVIEW

File: src/hooks/index.ts [hunk 1]
```diff
@@ -266,6 +269,89 @@ export const useCreateWallet = (options?: UseMutationOptions<WalletInterface, Er
   }
 }
 
+export const useSaveAndSubmitSignedTx = (
+  {wallet}: {wallet: WalletInterface},
+  options: UseMutationOptions<undefined | TxSubmissionStatus, Error, SignedTx> = {},
+) => {
+  const queryClient = useQueryClient()
+  const serverStatus = useSelector(serverStatusSelector)
```
Comment (wolverineks) [inline src/hooks/index.ts]: change to `await wallet.checkServerStatus()`

File: src/hooks/index.ts [hunk 2]
```diff
@@ -266,6 +269,89 @@ export const useCreateWallet = (options?: UseMutationOptions<WalletInterface, Er
...
+  const mutation = useMutation<undefined | TxSubmissionStatus, Error, SignedTx>({
```
Comment (wolverineks) [inline src/hooks/index.ts]: change to `useMutationWithInvalidations`

File: src/hooks/index.ts [hunk 3]
```diff
@@ -266,6 +269,89 @@ export const useCreateWallet = (options?: UseMutationOptions<WalletInterface, Er
...
+    mutationFn: async (signedTx: SignedTx) => {
+      queryClient.invalidateQueries([wallet.id, 'pendingTxs'])
+
+      await wallet.submitTransaction(signedTx.base64)
+
+      if (serverStatus.isQueueOnline) {
+        return await fetchTxStatus(wallet, signedTx.id, false)
```
Comment (wolverineks) [inline src/hooks/index.ts]: no await

File: src/hooks/index.ts [hunk 4]
```diff
@@ -266,6 +269,89 @@ export const useCreateWallet = (options?: UseMutationOptions<WalletInterface, Er
   }
 }
 
+export const useSaveAndSubmitSignedTx = (
```
Comment (wolverineks) [inline src/hooks/index.ts]: change to `useSubmitTx`

File: src/hooks/index.ts [hunk 5]
```diff
@@ -266,6 +269,89 @@ export const useCreateWallet = (options?: UseMutationOptions<WalletInterface, Er
...
+  }
+}
+
+const txQueueRetryDelay = 1000
+const txQueueRetryTimes = 5
+export const fetchTxStatus = async (
+  wallet: WalletInterface,
+  txHash: string,
+  waitProcessing = false,
+): Promise<TxSubmissionStatus | undefined> => {
```
Comment (wolverineks) [inline src/hooks/index.ts]: when does this return `undefined`?

File: src/hooks/index.ts [hunk 6]
```diff
@@ -266,6 +269,89 @@ export const useCreateWallet = (options?: UseMutationOptions<WalletInterface, Er
   }
 }
 
+export const useSaveAndSubmitSignedTx = (
+  {wallet}: {wallet: WalletInterface},
+  options: UseMutationOptions<undefined | TxSubmissionStatus, Error, SignedTx> = {},
```
Comment (wolverineks) [inline src/hooks/index.ts]: remove `undefined`

File: storybook/decorators/modals.tsx [hunk 7]
```diff
@@ -14,6 +14,7 @@ export const WithModalProps = ({children}: {children: childrenFn}) => {
 
   React.useEffect(() => {
     setTimeout(() => setVisible(true), 500) // weird behavior when starting with visible: true
+    setTimeout(() => setVisible(false), 3500) // auto-close
```
Comment (wolverineks) [inline storybook/decorators/modals.tsx:17]: change to `<Button onPress={() => setVisible(false)} title={'hide modal'} />`

File: src/components/PleaseWaitModal.tsx [hunk 8]
```diff
@@ -10,7 +11,9 @@ export const PleaseWaitView = ({title, spinnerText}: PleaseWaitViewProps) => (
   <View style={styles.container}>
     <Text style={styles.title}>{title}</Text>
 
-    <ActivityIndicator size="large" />
+    <ActivityIndicator size="large" color="#999999" />
```
Comment (wolverineks) [inline src/components/PleaseWaitModal.tsx]: check color

File: legacy/crypto/WalletInterface.js [hunk 9]
```diff
@@ -213,5 +215,7 @@ export interface WalletInterface {
 
   fetchFundInfo(): Promise<FundInfoResponse>;
 
+  fetchTxStatus(request: TxStatusRequest): Promise<TxStatusResponse>;
```
Comment (wolverineks) [inline legacy/crypto/WalletInterface.js:218]: is this the same as checkServerStatus?

File: legacy/crypto/shelley/ShelleyWallet.js [hunk 10]
```diff
@@ -882,4 +888,8 @@ export default class ShelleyWallet extends Wallet implements WalletInterface {
   async fetchFundInfo(): Promise<FundInfoResponse> {
     return await api.getFundInfo(this._getBackendConfig(), this._getNetworkConfig().IS_MAINNET)
   }
+
+  async fetchTxStatus(request: TxStatusRequest): Promise<TxStatusResponse> {
```
Comment (wolverineks) [inline legacy/crypto/shelley/ShelleyWallet.js:892]: `checkServerStatus`

File: src/Catalyst/Catalyst.hooks.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+export type CatalystData = {
+  catalystSKHex: string
+  catalystSKHexEncrypted: string
+  signRequest: HaskellShelleyTxSignRequest
+}
+export const useCatalyst = (
+  {wallet}: {wallet: WalletInterface},
+  options: UseMutationOptions<CatalystData, Error, CatalystVariables> = {},
+) => {
+  const serverStatus = useSelector(serverStatusSelector)
```
Comment (wolverineks) [inline src/Catalyst/Catalyst.hooks.ts]: `wallet.checkServerStatus()`

File: src/Catalyst/Catalyst.hooks.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+      return {
+        catalystSKHex,
+        catalystSKHexEncrypted,
+        signRequest,
+      }
+    },
+    ...options,
+  })
+
+  return mutation
```
Comment (wolverineks) [inline src/Catalyst/Catalyst.hooks.ts]: add aliasing

File: src/Catalyst/Step3.tsx [hunk 13]
```diff
@@ -3,31 +3,37 @@ import React, {useState} from 'react'
...
+}
+export const Step3 = ({pin, setCatalystData}: Props) => {
   const intl = useIntl()
   const strings = useStrings()
   const navigation = useNavigation()
-  const pin = useSelector(pinSelector)
-  const isHW = useSelector(isHWSelector)
-  const [confirmPin, setPin] = useState('')
+  const wallet = useSelectedWallet()
+  const {mutateAsync: generateVotingTransaction} = useCatalyst({wallet})
```
Comment (wolverineks) [inline src/Catalyst/Step3.tsx]: `mutate` not `mutateAsync`

File: src/Catalyst/Step4.tsx [hunk 14]
```diff
@@ -40,26 +43,30 @@ export const Step4 = () => {
...
         setGeneratingTransaction(false)
       }
       navigation.navigate(CATALYST_ROUTES.STEP5)
     }
 
-    if (isEasyConfirmationEnabled) {
+    if (wallet.isEasyConfirmationEnabled) {
       try {
-        await walletManager.ensureKeysValidity()
+        await ensureKeysValidity(walletManager._id)
```
Comment (wolverineks) [inline src/Catalyst/Step4.tsx]: wallet.id

File: src/Catalyst/Step3.tsx (hunk 13 above)
Comment (wolverineks) [inline src/Catalyst/Step3.tsx]: `const { data, refetch, error, isLoading } = useQuery(...)`

File: src/Catalyst/Step4.tsx [hunk 15]
```diff
@@ -24,14 +24,17 @@ type ErrorData = {
...
+  pin: string
+  setCatalystData: (catalystData?: CatalystData | undefined) => void
+}
+export const Step4 = ({pin, setCatalystData}: Props) => {
   const intl = useIntl()
   const strings = useStrings()
-  const isHW = useSelector(isHWSelector)
-  const isEasyConfirmationEnabled = useSelector(easyConfirmationSelector)
+  const wallet = useSelectedWallet()
+  const {mutateAsync: generateVotingTransaction} = useCatalyst({wallet})
```
Comment (wolverineks) [inline src/Catalyst/Step4.tsx]: `const { data, refetch, isLoading, error } = useQuery(...)`

File: src/Catalyst/Step6.tsx [hunk 16]
```diff
@@ -59,7 +60,10 @@ export const Step6 = () => {
       <ScrollView bounces={false} contentContainerStyle={styles.contentContainer}>
         <Spacer height={48} />
 
-        <Title>{strings.subTitle}</Title>
+        <Title>
+          {strings.subTitle}
+          {typeof catalystData?.signRequest}
```
Comment (wolverineks) [inline src/Catalyst/Step6.tsx]: remove

File: src/Dashboard/WithdrawStakingRewards/WithdrawStakingRewards.tsx [hunk 17]
```diff
@@ -225,7 +226,7 @@ export class WithdrawStakingRewards extends React.Component<Props, State> {
 
       if (isEasyConfirmationEnabled) {
         try {
-          await walletManager.ensureKeysValidity()
+          await ensureKeysValidity(walletManager._id)
```
Comment (wolverineks) [inline src/Dashboard/WithdrawStakingRewards/WithdrawStakingRewards.tsx:229]: wallet.id

**Review by wolverineks — APPROVED** [review]

File: src/Dashboard/WithdrawStakingRewards/WithdrawStakingRewards.tsx (hunk 17 above)
Comment (stackchain) [reply src/Dashboard/WithdrawStakingRewards/WithdrawStakingRewards.tsx:229]: here it is still a class, gonna be doing in your PR.

File: src/hooks/index.test.ts [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+      fetchTxStatus: Promise.resolve({
+        depth: {'12366dd454b4793d83d554329245d4bc67659f6f0468bb0e0d41f1ac6f496999': 1},
+      } as TxStatusResponse),
+    })
+
+    // act
+    const result = await fetchTxStatus(wallet, txId)
+
+    // assert
+    expect(wallet.fetchTxStatus.alwaysCalledWith(requestShouldBe)).to.be.true
```
Comment (wolverineks) [inline src/hooks/index.test.ts:26]: this is testing implementation details

File: src/hooks/index.test.ts [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+    expect(wallet.fetchTxStatus.callCount).to.be.equal(1)
+    expect(result).to.be.eql({
```
Comment (wolverineks) [inline src/hooks/index.test.ts:28]: this is the only test that matters

File: src/hooks/index.test.ts [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+
+describe('fetchTxStatus', () => {
+  // it means that the tx was sent to node and processed
+  it('should return success when depth > 0', async () => {
+    // arrange
+    const txId = '12366dd454b4793d83d554329245d4bc67659f6f0468bb0e0d41f1ac6f496999'
+    const requestShouldBe: TxStatusRequest = {
+      txHashes: [txId],
+    }
+    const wallet = sinon.stubObject(walletFactory({}), {
```
Comment (wolverineks) [inline src/hooks/index.test.ts:16]: sinon is not necessary
stubbing/spying is not necessary
only the return value of fetchTxStatus matters

Comment (stackchain) [reply src/hooks/index.test.ts:16]: I'm using stub so I can get the benefits of it, like the call count to ensure that it was retrying.

**Pull request merged**